test/components/linkerd/linkerd_test.go

```diff
@@ -0,0 +1,230 @@
+    command:
+    - bash
+    args:
+    - -c
+    - 'curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x ./kubectl && cp ./kubectl /data/'
+    volumeMounts:
+    - name: download-dir
+      mountPath: /data
+  - image: fedora
+    name: download-linkerd
+    command:
+    - bash
+    args:
+    - -c
+    - 'curl -L https://run.linkerd.io/install | sh && cp -L /root/.linkerd2/bin/linkerd /data/linkerd'
+    volumeMounts:
+    - name: download-dir
+      mountPath: /data
+  containers:
+  - image: fedora
+    name: linkerd-cli
+    env:
+    - name: KUBECONFIG
+      value: /root/.kube/config
+    command:
+    - bash
+    args:
+    - -c
+    - 'mv /data/* /usr/local/bin/ && kubectl version -o yaml && linkerd version && linkerd check'
+    volumeMounts:
+    - name: kubeconfig
+      mountPath: /root/.kube
+    - name: download-dir
+      mountPath: /data
+  volumes:
+  - name: kubeconfig
+    secret:
+      secretName: kubeconfig
+  - name: download-dir
+    emptyDir: {}
+`
+
+	retryInterval = 5 * time.Second
+	timeout       = 9 * time.Minute
+)
+
+// nolint: funlen
+func TestLinkerdCheck(t *testing.T) {
+	t.Parallel()
+
+	client := testutil.CreateKubeClient(t).CoreV1()
+	nsclient := client.Namespaces()
+
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: testutil.TestNamespace("linkerd-check-test"),
+		},
+	}
+
+	// Parsing.
+	pod := &corev1.Pod{}
+	if err := yaml.Unmarshal([]byte(podConfig), pod); err != nil {
+		t.Fatalf("failed to unmarshal pod manifest: %v", err)
+	}
+
+	// Creation.
+	ns, err := nsclient.Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("failed creating Namespace: %v", err)
+	}
+
+	t.Cleanup(func() {
+		if err := nsclient.Delete(context.TODO(), ns.Name, metav1.DeleteOptions{}); err != nil {
+			t.Logf("failed removing Namespace: %v", err)
+		}
+	})
+
+	_, err = client.Secrets(ns.Name).Create(context.TODO(), getKubeconfigSecret(t), metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("failed creating Secret: %v", err)
+	}
+
+	podsClient := client.Pods(ns.Name)
+
+	// Retry pod creation. This might fail if Linkerd is not ready yet and some requests might fail.
+	if err := wait.PollImmediate(retryInterval, timeout, func() (done bool, err error) {
+		pod, err = podsClient.Create(context.TODO(), pod, metav1.CreateOptions{})
+		if err != nil {
+			t.Logf("retrying pod creation, failed with: %v", err)
+
+			return false, nil
+		}
+
+		return true, nil
+	}); err != nil {
+		t.Fatalf("error while trying to create the pod: %v", err)
+	}
+
+	phase := corev1.PodUnknown
+
+	if err := wait.PollImmediate(retryInterval, timeout, func() (done bool, err error) {
+		p, err := podsClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("couldn't get the pod %q: %w", pod.Name, err)
```

Comment:
Should we log and retry here if e.g. API server is flaky?

Reply:
This can be a lot of output. Hence I avoided it. Also with timeout we will get the same error, which will be logged anyway.

Reply:
Fair enough :+1: